pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -81,27 +85,75 @@ func PodChanged(oldPod, newPod *v1.Pod, endpointChanged EndpointsMatch) (bool, b
 
 // GetPodServiceMemberships returns a set of Service keys for Services that have
 // a selector matching the given pod.
-func GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
+func GetPodServiceMemberships(serviceLister v1listers.ServiceLister, selectorCache *sync.Map, pod *v1.Pod) (sets.String, error) {
 	set := sets.String{}
-	services, err := serviceLister.GetPodServices(pod)
+	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
 	if err != nil {
 		// don't log this error because this function makes pointless
 		// errors when no services match
 		return set, nil
 	}
-	for i := range services {
-		key, err := controller.KeyFunc(services[i])
-		if err != nil {
-			return nil, err
+
+	var selector labels.Selector
+	for _, service := range services {
+		if v, ok := selectorCache.Load(fmt.Sprintf("%s_%s", service.Namespace, service.Name)); ok {
+			selector = v.(labels.Selector)
+		} else {
+			if service.Spec.Selector == nil {
+				// if the service has a nil selector this means selectors match nothing, not everything.
+				continue
+			}
+			selector = labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+			selectorCache.Store(fmt.Sprintf("%s_%s", service.Namespace, service.Name), selector)
+		}
+		if selector.Matches(labels.Set(pod.Labels)) {
+			key, err := controller.KeyFunc(service)
+			if err != nil {
+				return nil, err
+			}
+			set.Insert(key)
 		}
-		set.Insert(key)
 	}
 	return set, nil
 }
 
+// SyncSelectorCache will confirm consistency between local selectorCache and service's selector.
+func SyncSelectorCache(service *v1.Service, selectorCache *sync.Map) {
+	v, ok := selectorCache.Load(fmt.Sprintf("%s_%s", service.Namespace, service.Name))
```

Comment:
I may be misunderstanding how this works, but it seems like you could end up with a race condition where the cache is modified between the load and set here. I think it might be better and safer to use a standard map with an associated lock that you can use to lock the whole function between read and write. There are some relevant examples of this pattern [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpointslicecache.go#L170), though probably also better examples elsewhere. This would also bring back type safety here and allow you to use a map like `map[types.NamespacedName]labels.Selector` for simplicity.